### PR TITLE
export repository with support for transactions [MARXAN-871] [MARXAN-1174]

### PR DESCRIPTION
**Pull request by AaronPerezPerez**
### Overview

Implements the Export Repository for TypeORM with a find and save method. Also, this PR adds tests for the repository, and finally we also added a transaction method to do transactional tasks.

## Related issues
[MARXAN-1147 Export Repository should support transactions](https://vizzuality.atlassian.net/browse/MARXAN-1147)
[MARXAN-871 Export Repository](https://vizzuality.atlassian.net/browse/MARXAN-871)

### CODE REVIEW

**Review by hotzevzl — CHANGES_REQUESTED** [review]
@aciddaute @AaronPerezPerez @angelhigueraacid good job!

The implementation looks good to me, I only recommend to adjust the test fixtures to stick to clear step separation (should be a fairly minor change).

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts [hunk 1]
```diff
@@ -179,5 +194,15 @@ const getFixtures = async () => {
...
     },
+    ThenAllExportComponentsShouldBeFinished: ({
+      exportData,
+    }: {
+      exportData: Export | undefined;
+    }) => {
+      expect(exportData).toBeDefined();
+      exportData!.toSnapshot().exportPieces.map((piece) => {
+        expect(piece.finished).toEqual(true);
+      });
```
Comment (kgajowy) [inline api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:205]: assertions in mapping look really strange ; )

`[{finished: true}, {finished: false}].every(i => i.finished)`

Comment (aciddaute) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:205]: I'll fix it with a `forEach` :+1:

File: api/apps/api/src/modules/clone/export/adapters/entities/exports.api.entity.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  static fromAggregate(exportAggregate: Export): ExportEntity {
+    const snapshot = exportAggregate.toSnapshot();
+
+    const exportEntity = new ExportEntity();
+
+    exportEntity.id = snapshot.id;
+    exportEntity.resourceId = snapshot.resourceId;
+    exportEntity.resourceKind = snapshot.resourceKind;
+    exportEntity.archiveLocation = snapshot.archiveLocation;
+    exportEntity.components = snapshot.exportPieces.map((piece) => {
```
Comment (kgajowy) [inline api/apps/api/src/modules/clone/export/adapters/entities/exports.api.entity.ts:43]: also consider `class-transformer` to avoid too much typing

File: api/apps/api/src/modules/clone/export/adapters/typeorm-export.repository.ts [hunk 3]
```diff
@@ -1,16 +1,63 @@
...
+      resourceKind: exportEntity.resourceKind,
+      archiveLocation: exportEntity.archiveLocation,
+      exportPieces: exportEntity.components.map((component) => ({
+        id: new ComponentId(component.id),
+        finished: component.finished,
+        piece: component.piece,
+        resourceId: component.resourceId,
+        uris: component.uris.map(
+          (element) => new ComponentLocation(element.uri, element.relativePath),
+        ),
```
Comment (kgajowy) [inline api/apps/api/src/modules/clone/export/adapters/typeorm-export.repository.ts:42]: Looks like mapping knowledge is now spread to both repo and entity. Consider putting it in one place.

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+  it('should update an export when a nested piece is completed', async () => {
+    await fixtures.WhenAnExportIsSavedSuccessfully();
+    await fixtures.WhenAComponentIsCompleted();
+    const exportData = await fixtures.WhenReadingTheSavedExportFromRepository();
+
+    await fixtures.ThenExportDataShouldBeOk({
+      exportData,
+      componentsAreCompleted: true,
+    });
+  });
```
Comment (kgajowy) [inline api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:45]: this may be a bit confusing. Test says `should update an export when a nested piece is complete` so I would rather expect more process/product naming and 'events' what actually happen.

```
GivenExportWasRequested();
WhenLastPieceWasCompleted();
ThenExportIsSaved();
ThenExportFinalizedEventIsSent();  // don't remember the name, sorry
```

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+  it('should save entities working with transaction', async () => {
+    await fixtures.WhenAnExportIsSavedSuccessfullyWithinATransaction();
+    await fixtures.WhenAllComponentsAreCompletedWithTransactions();
+
+    const exportData = await fixtures.WhenReadingTheSavedExportFromRepository();
+
+    fixtures.ThenAllExportComponentsShouldBeFinished({
+      exportData,
+    });
```
Comment (kgajowy) [inline api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:55]: `withTransactions` seems to be irrelevant technical detail there 🤔

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts (hunk 1 above)
Comment (kgajowy) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:205]: @aciddaute but why `forEach`, if you can do that with `every()` ?

Comment (aciddaute) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:205]: Oh sorry, I didn't see that, I'll use `every()` ty Kamil 🙌

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts (hunk 4 above)
Comment (AaronPerezPerez) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:45]: As this is an integration test for our custom repository, this test should not know about events. Also, the second step is just one completed piece, not every one / the last one

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts (hunk 5 above)
Comment (AaronPerezPerez) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:55]: As we are testing the repository, we wanted to have another integrtation test for that specific method of our repository

Comment (kgajowy) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts]: @AaronPerezPerez hm, steps deceived me. How about then testing it truly as an integration?

```
export = GivenTheExport()
WhenSaving(fixtures.completePieceA(export);
WhenSaving(fixtures.completePieceB(export);
ThenBothPiecesAreSaved(); // just use entity => export like in normal application and verify on "domain" object.
```

File: api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts [hunk 6]
```diff
@@ -6,15 +6,25 @@ import { ExportRepository } from '../application/export-repository.port';
 import { ResourcePieces } from '../application/resource-pieces.port';
 import { ArchiveCreator } from '../application/archive-creator.port';
 import { ResourcePiecesAdapter } from './resource-pieces.adapter';
-import { InMemoryExportRepo } from './in-memory-export.repository';
 import { NodeArchiveCreator } from './node-archive-creator';
+import { TypeormExportRepository } from '@marxan-api/modules/clone/export/adapters/typeorm-export.repository';
+import { ExportEntity } from '@marxan-api/modules/clone/export/adapters/entities/exports.api.entity';
+import { ExportComponentEntity } from '@marxan-api/modules/clone/export/adapters/entities/export-components.api.entity';
+import { ComponentLocationEntity } from '@marxan-api/modules/clone/export/adapters/entities/component-locations.api.entity';
```
Comment (kgajowy) [inline api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts:13]: relative imports (:

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+import { ClonePiece, ComponentId, ResourceKind } from '@marxan/cloning/domain';
+import { FixtureType } from '@marxan/utils/tests/fixture-type';
+import { Connection } from 'typeorm';
+import { v4 } from 'uuid';
+import { ResourceId } from '../../src/modules/clone/export';
+import { ExportEntity } from '../../src/modules/clone/export/adapters/entities/exports.api.entity';
+import { ComponentLocation } from '../../src/modules/clone/export/application/complete-piece.command';
+import { ExportRepository } from '../../src/modules/clone/export/application/export-repository.port';
+import { Export, ExportId } from '../../src/modules/clone/export/domain';
+import { bootstrapApplication } from '../utils/api-application';
```
Comment (kgajowy) [inline api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:12]: @paths there on the other hand

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+      const connection = app.get<Connection>(Connection);
+      const exportRepo = connection.getRepository(ExportEntity);
+      const exportComponentRepo = connection.getRepository(
+        ExportComponentEntity,
+      );
+      const locationRepo = connection.getRepository(ComponentLocationEntity);
+
+      await locationRepo.delete({});
+      await exportComponentRepo.delete({});
+      await exportRepo.delete({});
```
Comment (kgajowy) [inline api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts:80]: could be a good secret test of cascade delete (running only `await exportRepo.delete({});`)

File: api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts (hunk 5 above)
Comment (aciddaute) [reply api/apps/api/test/integration/typeorm-export.repository.integration.e2e-spec.ts]: But if we want to test the `transaction` method behaviour we should complete several pieces concurrently 🤔 . The steps of the current test are:

1. `GivenExportWithMultipleComponentsWasRequested`: Requests an export with 10 pieces
2. `WhenAllComponentsAreCompletedConcurrentlyWithTransactions`: Completes those pieces concurrently using `transaction` method
3. `WhenReadingTheSavedExportFromRepository`: Reads the export from the repo
4. `ThenAllExportComponentsShouldBeFinished`: Checks that all pieces are marked as finished

Note that the code above is outdated.

File: api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts (hunk 6 above)
Comment (tiagojsag) [reply api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts:13]: Do we have a documented rule of when to use absolute/relative imports? I've seen multiple comments on PRs around this, so I'm wondering if everyone is aware of exactly how we should approach this - I am not, for one.

File: api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts [hunk 9]
```diff
@@ -6,15 +6,25 @@ import { ExportRepository } from '../application/export-repository.port';
...
+    TypeOrmModule.forFeature([
+      ExportEntity,
+      ExportComponentEntity,
+      ComponentLocationEntity,
+    ]),
+  ],
   providers: [
     {
       provide: ExportRepository,
-      useClass: InMemoryExportRepo, // TODO TypeormExportRepository
```
Comment (tiagojsag) [inline api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts:17]: This seems like the last usage of the InMemory repo besides tests, so we should probably get rid of it.
A related question: why is ExportRepository an abstract class instead of an interface?

Comment (aciddaute) [reply api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts:17]: In order to be able to use `ExportRepository` as the `provide` property of the provider. We can't use a Typescript interface as a value for `provide` property. We could also declare it as an interface and use a Symbol for the `provide` property along as we use the `@Inject` decorator for DI. `TypeormExportRepository` and `InMemoryExportRepo` implements the abstract class, so we should be safe 🤔

**Review by hotzevzl — APPROVED** [review]

File: api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts (hunk 6 above)
Comment (hotzevzl) [reply api/apps/api/src/modules/clone/export/adapters/export-adapters.module.ts]: we don't, and I got lost in the evolution of this through time.

My preference would be:
- within a Nest module, relative paths
- across modules within the same service, "root-relative" paths
- across services or from libs, `@<whatever>` "root-relative" paths

though IDEs should make the difference between the first two cases mostly cosmetic - personally I like "root-relative" paths as I find that it removes the small cognitive overload that (at least for me) relative paths introduce by having to remember or figure out "where am I" in the paths hierarchy to then travel up and down the tree mentally. But even this, I normally deal with ok by just ctrl-clicking through the relevant imports if I can't figure them out at a glance.

**Review by tiagojsag — APPROVED** [review]
